### ISSUE TITLE
[SB] Poor desktop and render performance

### ISSUE THREAD

**Issue by chiekku** [linux-surface/linux-surface#213]
This is mainly me trying to understand the current state of performance and graphics rendering with the kernel. I've been enjoying using Linux on this device a lot more than Windows 10 (which has gotten worse over time as far as I can tell). However, rendering performance has been buggy for me, so I want to try and figure out why. I'm curious what other users are seeing, too.

This will be very long, so I'll put a TL;DR at the end.

By default the system takes forever to do anything. Since this is Skylake, Ubuntu is using the ``intel_pstate`` module; the governer defaults to ``Powersave``. To fix this I followed the steps [here](https://askubuntu.com/a/936488).

Now I can set the governer to ``Performance``, which improved the overall snappy-ness of the system. Definitely recommend. 

It's definitely better, but it still lags when doing menial tasks like opening windows (e.g. 3-5 seconds for a calculator window to pop up). This is hard to quantify other than how it feels.

Geekbench 5: https://browser.geekbench.com/v5/cpu/2577333
```
	892 Single-Core
	1945 Multi-Core
```
Looking thru the other results for this model, it looks like mine scored nominally higher than average, and most of them are Windows 10. So that's cool! But rendering still stutters, so I guess I need to dive deeper.

For some basic vga info, I ran these:
```
	$ glxinfo -B
	name of display: :0
	display: :0  screen: 0
	direct rendering: Yes
	Extended renderer info (GLX_MESA_query_renderer):
	    Vendor: Intel (0x8086)
	    Device: Mesa Intel(R) HD Graphics 520 (SKL GT2) (0x1916)
	    Version: 20.2.0
	    Accelerated: yes
	    Video memory: 3072MB
	    Unified memory: yes
	    Preferred profile: core (0x1)
	    Max core profile version: 4.6
	    Max compat profile version: 4.6
	    Max GLES1 profile version: 1.1
	    Max GLES[23] profile version: 3.2
	OpenGL vendor string: Intel
	OpenGL renderer string: Mesa Intel(R) HD Graphics 520 (SKL GT2)
	OpenGL core profile version string: 4.6 (Core Profile) Mesa 20.2.0-devel (git-4de678c 2020-06-13 eoan-oibaf-ppa)
	OpenGL core profile shading language version string: 4.60
	OpenGL core profile context flags: (none)
	OpenGL core profile profile mask: core profile

	OpenGL version string: 4.6 (Compatibility Profile) Mesa 20.2.0-devel (git-4de678c 2020-06-13 eoan-oibaf-ppa)
	OpenGL shading language version string: 4.60
	OpenGL context flags: (none)
	OpenGL profile mask: compatibility profile

	OpenGL ES profile version string: OpenGL ES 3.2 Mesa 20.2.0-devel (git-4de678c 2020-06-13 eoan-oibaf-ppa)
	OpenGL ES profile shading language version string: OpenGL ES GLSL ES 3.20
```
This is using the latest xorg intel driver; so far so good.
```
	$ sudo lspci -v -s 00:02.0
	00:02.0 VGA compatible controller: Intel Corporation Skylake GT2 [HD Graphics 520] (rev 07) (prog-if 00 [VGA controller])
		Subsystem: Microsoft Corporation Skylake GT2 [HD Graphics 520]
		Flags: bus master, fast devsel, latency 0, IRQ 131
		Memory at a0000000 (64-bit, non-prefetchable) [size=16M]
		Memory at 90000000 (64-bit, prefetchable) [size=256M]
		I/O ports at 8000 [size=64]
		[virtual] Expansion ROM at 000c0000 [disabled] [size=128K]
		Capabilities: [40] Vendor Specific Information: Len=0c <?>
		Capabilities: [70] Express Root Complex Integrated Endpoint, MSI 00
		Capabilities: [ac] MSI: Enable+ Count=1/1 Maskable- 64bit-
		Capabilities: [d0] Power Management version 2
		Capabilities: [100] Process Address Space ID (PASID)
		Capabilities: [200] Address Translation Service (ATS)
		Capabilities: [300] Page Request Interface (PRI)
		Kernel driver in use: i915
		Kernel modules: i915
```
Okay, so If i understand this correctly, we have: 256 MB + 16 MB dedicated memory, and some 2800 MB shared system memory for our total of 3072 MB unified video memory. That sounds about right from what I remember seeing in Windows. 

Now for some benchmarks.

glxgears runs fine, so we can confirm basic functionality:
	``300 frames in 5.0 seconds = 59.974 FPS``

glmark2 runs fine; most tests run above 1000 fps, but all tests exhibit stuttering or hitching of some kind. 

Unigine Heaven 4.0 (2013) benchmark on lowest possible settings, 720p resolution:	
```
        Score: 871
	Min FPS: 8.2
	Avg FPS: 34.6
	Max FPS: 62.1
```
It's pretty bad. There's significant stutter/hitching here, which is why the range of values is so large and why the lowest recorded FPS value is 8.

It also has issues identifying the GPU; from the log:
```
---- System ----
System: Linux 5.7.2-surface x86_64
CPU: Intel(R) Core(TM) i7-6600U CPU @ 2.60GHz 2807MHz MMX SSE SSE2 SSE3 SSSE3 SSE41 SSE42 AVX HTT x4
GPU: Unknown GPU x1
System memory: 15935 MB
Video memory:  256 MB
```
Wait a minute... why is it not able to tell what GPU this is? It's also not reporting the shared/unified memory. Are we maybe running out of VRAM?

Unigine Superposition 1.1 (2019) however correctly identifies the 520 HD with 3072 MB. On 720p low, it claims to use 643 MB of VRAM:
```
        Score: 1756
	FPS Min: 9.79
	FPS Avg: 13.14
	FPS Max: 17.38
```

Obviously it runs at lower avg FPS, but it runs "smoothly". Score is higher, framerates are more consistent, no noticeable stutter like in Heaven. Amusingly, it has a nominally higher min FPS value despite being a more intense load. There's something different between how the two programs are identifying/using the iGPU.

**TL;DR: In addition to laggy desktop performance, most video rendering suffers from significant stutter. It appears there's either issues with shared/unified memory, or how iGPU capabilities are being reported.**

Please post any benchmarks you have done so we can all compare notes! I'll update this post with more as I find it.
